extension/json/json_functions/json_structure.cpp

```diff
@@ -0,0 +1,186 @@
+	yyjson_mut_val *key, *val;
+	yyjson_mut_obj_iter iter;
+	vector<string> key_insert_order;
+	unordered_map<string, vector<yyjson_mut_val *>> key_values;
+	for (const auto &elem_structure : elem_structures) {
+		yyjson_mut_obj_iter_init(elem_structure, &iter);
+		while ((key = yyjson_mut_obj_iter_next(&iter))) {
+			auto key_string = yyjson_mut_get_str(key);
+			val = yyjson_mut_obj_iter_get_val(key);
+			if (key_values.find(key_string) == key_values.end()) {
+				key_insert_order.push_back(key_string);
+			}
+			key_values[key_string].push_back(val);
+		}
+	}
+	auto result = yyjson_mut_obj(structure_doc);
+	for (const auto &key_string : key_insert_order) {
+		key = yyjson_mut_strncpy(structure_doc, key_string.c_str(), key_string.size());
+		val = GetConsistentArrayStructure(key_values.at(key_string), structure_doc);
+		yyjson_mut_obj_add(result, key, val);
+	}
+	return result;
+}
+
+static inline bool StringEquals(const char *lhs, const char *rhs) {
+	return strcmp(lhs, rhs) == 0;
+}
+
+static inline bool EitherEquals(const char *s1, const char *s2, const char *target) {
+	return StringEquals(s1, target) || StringEquals(s2, target);
+}
+
+static inline void GetMaxTypeString(const char *type_string, const char *elem_type_string, const char **result) {
+	if (!type_string) {
+		*result = elem_type_string;
+	} else if (!elem_type_string) {
+		*result = type_string;
+	} else if (StringEquals(type_string, elem_type_string)) {
+		*result = type_string;
+	} else if (EitherEquals(type_string, elem_type_string, JSONCommon::TYPE_STRING_VARCHAR)) {
+		*result = JSONCommon::TYPE_STRING_VARCHAR;
+	} else if (EitherEquals(type_string, elem_type_string, JSONCommon::TYPE_STRING_DOUBLE)) {
+		*result = JSONCommon::TYPE_STRING_DOUBLE;
+	} else if (EitherEquals(type_string, elem_type_string, JSONCommon::TYPE_STRING_BIGINT)) {
+		*result = JSONCommon::TYPE_STRING_BIGINT;
+	} else if (EitherEquals(type_string, elem_type_string, JSONCommon::TYPE_STRING_UBIGINT)) {
+		*result = JSONCommon::TYPE_STRING_UBIGINT;
+	} else {
+		D_ASSERT(StringEquals(type_string, JSONCommon::TYPE_STRING_BOOLEAN) &&
+		         StringEquals(elem_type_string, JSONCommon::TYPE_STRING_BOOLEAN));
+		*result = JSONCommon::TYPE_STRING_BOOLEAN;
+	}
+}
+
+static inline yyjson_mut_val *GetConsistentArrayStructure(const vector<yyjson_mut_val *> &elem_structures,
+                                                          yyjson_mut_doc *structure_doc) {
+	if (elem_structures.empty()) {
+		return yyjson_mut_str(structure_doc, JSONCommon::TYPE_STRING_NULL);
+	}
+	auto type = yyjson_mut_get_type(elem_structures[0]);
+	auto type_string = yyjson_mut_get_str(elem_structures[0]);
+	for (idx_t i = 1; i < elem_structures.size(); i++) {
+		auto elem_type = yyjson_mut_get_type(elem_structures[i]);
+		auto elem_type_string = yyjson_mut_get_str(elem_structures[i]);
+		if (type_string && strcmp(type_string, JSONCommon::TYPE_STRING_NULL) == 0) {
```

Comment:
Do we need these string operations for type comparisons, or can we do this directly on the enums?